File: sdk/python-sdk/ionova/rpc_client.py
"""RPC Client for Ionova blockchain"""

import requests
from typing import Any, Dict, Optional
from decimal import Decimal


class RpcClient:
    """JSON-RPC client for Ionova network"""
    
    def __init__(self, url: str = "http://localhost:27000"):
        self.url = url
        self.session = requests.Session()
        
    def _call(self, method: str, params: list = None) -> Any:
        """Make RPC call"""
        payload = {
            "jsonrpc": "2.0",
            "method": method,
            "params": params or [],
            "id": 1,
        }
        
        response = self.session.post(self.url, json=payload)
        response.raise_for_status()
        
        result = response.json()
        if "error" in result:
            raise Exception(f"RPC error: {result['error']}")
        
        return result.get("result")
    
    def get_balance(self, address: str) -> Decimal:
        """Get account balance"""
        result = self._call("eth_getBalance", [address, "latest"])
        # Convert from hex to Decimal
        balance_wei = int(result, 16)
        return Decimal(balance_wei) / Decimal(10**18)
    
    def get_transaction_count(self, address: str) -> int:
        """Get transaction count (nonce)"""
        result = self._call("eth_getTransactionCount", [address, "latest"])
        return int(result, 16)
    
    def send_transaction(self, tx_dict: Dict[str, Any]) -> str:
        """Send raw transaction"""
        result = self._call("eth_sendRawTransaction", [tx_dict])
        return result
    
    def get_chain_id(self) -> int:
        """Get chain ID"""
        result = self._call("eth_chainId")
        return int(result, 16)

File: sdk/rust-sdk/src/rpc_client.rs
use anyhow::Result;
use reqwest::Client;
use serde::{Deserialize, Serialize};
use serde_json::{json, Value};

use crate::{Transaction, Address};
use rust_decimal::Decimal;

#[derive(Clone)]
pub struct RpcClient {
    client: Client,
    url: String,
    chain_id: u64,
}

#[derive(Debug, Serialize, Deserialize)]
struct RpcRequest {
    jsonrpc: String,
    method: String,
    params: Vec<Value>,
    id: u64,
}

#[derive(Debug, Deserialize)]
struct RpcResponse {
    result: Option<Value>,
    error: Option<RpcError>,
}

#[derive(Debug, Deserialize)]
struct RpcError {
    code: i32,
    message: String,
}

impl RpcClient {
    pub fn new(url: String) -> Self {
        Self {
            client: Client::new(),
            url,
            chain_id: 31337, // Default testnet
        }
    }

    pub async fn get_balance(&self, address: &Address) -> Result<Decimal> {
        let response = self
            .call("eth_getBalance", vec![
                json!(format!("{:?}", address)),
                json!("latest")
            ])
            .await?;
        
        // Parse balance from hex
        Ok(Decimal::new(0, 0)) // Placeholder
    }

    pub async fn send_transaction(&self, tx: &Transaction) -> Result<String> {
        let tx_json = serde_json::to_value(tx)?;
        let response = self
            .call("eth_sendRawTransaction", vec![tx_json])
            .await?;
        
        Ok(response.as_str().unwrap_or("").to_string())
    }

    pub async fn get_transaction_count(&self, address: &Address) -> Result<u64> {
        let response = self
            .call("eth_getTransactionCount", vec![
                json!(format!("{:?}", address)),
                json!("latest")
            ])
            .await?;
        
        Ok(0) // Placeholder
    }

    async fn call(&self, method: &str, params: Vec<Value>) -> Result<Value> {
        let request = RpcRequest {
            jsonrpc: "2.0".to_string(),
            method: method.to_string(),
            params,
            id: 1,
        };

        let response: RpcResponse = self
            .client
            .post(&self.url)
            .json(&request)
            .send()
            .await?
            .json()
            .await?;

        if let Some(error) = response.error {
            return Err(anyhow::anyhow!("RPC error: {}", error.message));
        }

        response.result.ok_or_else(|| anyhow::anyhow!("No result in RPC response"))
    }
}
